src/cluster_asm.c

```diff
@@ -129,24 +146,91 @@ char *asmTaskStateToString(int state) {
+        case ASM_RDBCHANNEL_DONE: return "rdbchannel-done";
+
         default: return "unknown";
     }
     return NULL; /* Unreachable */
 }
 
+int asmTaskSetFailPoint(sds channel, sds state) {
+    if (!asmManager) {
+        serverLog(LL_WARNING, "ASM manager is not initialized");
+        return C_ERR;
+    }
+    asmManager->failed_channel = 0;
+    asmManager->failed_state = 0;
+    if (!channel && !state) return C_ERR;
+    if (sdslen(channel) == 0 && sdslen(state) == 0) {
+        serverLog(LL_WARNING, "ASM fail point is cleared");
+        return C_OK;
+    }
+
+    if (!strcasecmp(channel, "import-main-channel")) asmManager->failed_channel = ASM_IMPORT_MAIN_CHANNEL;
+    else if (!strcasecmp(channel, "import-rdb-channel")) asmManager->failed_channel = ASM_IMPORT_RDB_CHANNEL;
+    else if (!strcasecmp(channel, "migrate-main-channel")) asmManager->failed_channel = ASM_MIGRATE_MAIN_CHANNEL;
+    else if (!strcasecmp(channel, "migrate-rdb-channel")) asmManager->failed_channel = ASM_MIGRATE_RDB_CHANNEL;
+    else return C_ERR;
+
+    if (!strcasecmp(state, "connecting")) asmManager->failed_state = ASM_CONNECTING;
+    else if (!strcasecmp(state, "auth-reply")) asmManager->failed_state = ASM_AUTH_REPLY;
+    else if (!strcasecmp(state, "handshake-reply")) asmManager->failed_state = ASM_HANDSHAKE_REPLY;
+    else if (!strcasecmp(state, "syncslots-reply")) asmManager->failed_state = ASM_SYNCSLOTS_REPLY;
+    else if (!strcasecmp(state, "accumulate-buffer")) asmManager->failed_state = ASM_ACCUMULATE_BUF;
+    else if (!strcasecmp(state, "streaming-buffer")) asmManager->failed_state = ASM_STREAMING_BUF;
+    else if (!strcasecmp(state, "wait-stream-eof")) asmManager->failed_state = ASM_WAIT_STREAM_EOF;
+    else if (!strcasecmp(state, "wait-rdbchannel")) asmManager->failed_state = ASM_WAIT_RDBCHANNEL;
+    else if (!strcasecmp(state, "wait-bgsave-start")) asmManager->failed_state = ASM_WAIT_BGSAVE_START;
+    else if (!strcasecmp(state, "send-bulk-and-stream")) asmManager->failed_state = ASM_SEND_BULK_AND_STREAM;
+    else if (!strcasecmp(state, "paused-write")) asmManager->failed_state = ASM_PAUSED_WRITE;
+    else if (!strcasecmp(state, "rdbchannel-reply")) asmManager->failed_state = ASM_RDBCHANNEL_REPLY;
+    else if (!strcasecmp(state, "rdbchannel-transfer")) asmManager->failed_state = ASM_RDBCHANNEL_TRANSFER;
+    else return C_ERR;
+
+    serverLog(LL_NOTICE, "ASM fail point set: channel=%s, state=%s", channel, state);
+    return C_OK;
+}
+
+const char *asmChannelToString(int channel) {
+    switch (channel) {
+        case ASM_IMPORT_MAIN_CHANNEL: return "import-main-channel";
+        case ASM_IMPORT_RDB_CHANNEL: return "import-rdb-channel";
+        case ASM_MIGRATE_MAIN_CHANNEL: return "migrate-main-channel";
+        case ASM_MIGRATE_RDB_CHANNEL: return "migrate-rdb-channel";
+        default: return "unknown";
+    }
+}
+
+int asmIsFailPointActive(int channel, int state) {
```

Comment:
perhaps we should use "debug" word somewhere in variable names and function names as much as possible to make it obvious what we are doing. e.g. `asmDebugIsFailPointActive()`. Otherwise, I feel like it won't be immediately obvious what we are doing. It can be confused with failover etc.

Reply:
make sense